niceman/utils.py

```diff
@@ -895,4 +896,21 @@ def unicode(s):
     else:
         return __builtin__.unicode(s)
 
+
+def generate_unique_name(pattern, nameset):
+    """Create a unique numbered name from a pattern and a set"""
+    i = 0
+    while True:
+        n = pattern % i
+        i += 1
+        if n not in nameset:
+            return n
+
+
+# http://stackoverflow.com/questions/1151658/python-hashable-dicts
+class HashableDict(dict):
+    """Dict that can be used as keys"""
+    def __hash__(self):
+        return hash(frozenset(self.values()))
```

Comment:
looking back into the code -- this HashableDict is used just to collate those origins into their 'set'.  so you could have just got a 'get_dict_hash' function, thus avoiding giving a promise on any kind of immutable hashable dict, and then use while deducing those unique origins.  anyways that dictionary is not returned since you only care about values at the end (actual "Origin"s)

and I think we should come back to our discussion on using `attr` module to provide specifications for all those constructs (for Origin, etc).  dicts are nice but now we have no clearly defined set of fields to expect in those etc.  Then they would not be dicts at all and we would need just to adjust/create one function instead of some "improved immutable class" ;)